app/liquid_tags/unified_embed/tag.rb

```diff
@@ -24,43 +25,43 @@ class Tag < LiquidTagBase
     def self.new(tag_name, input, parse_context)
       stripped_input = ActionController::Base.helpers.strip_tags(input).strip
 
-      # Extract just the URL from the input, without any params, for validation
-      actual_link = extract_only_url(stripped_input)
-
-      # When Listings are disabled, it makes little sense to perform a validate_link
+      # when Listings are disabled, it makes little sense to perform a validate_link
       # network call.
-      handle_listings_disabled!(actual_link)
+      handle_listings_disabled!(stripped_input)
 
       # Before matching against the embed registry, we check if the link
       # is valid (e.g. no typos).
       # If the link is invalid, we raise an error encouraging the user to
       # check their link and try again.
-      validate_link!(actual_link)
-      klass = UnifiedEmbed::Registry.find_liquid_tag_for(link: stripped_input)
+      validated_link = validate_link(input: stripped_input)
+      klass = UnifiedEmbed::Registry.find_liquid_tag_for(link: validated_link)
 
-      # If there are no OG metatags, we shall render an A-tag. Since the link
-      # has been validated, at least this A-tag will not 404.
-      #
       # Why the __send__?  Because a LiquidTagBase class "privatizes"
       # the `.new` method.  And we want to instantiate the specific
       # liquid tag for the given link.
-      klass.__send__(:new, tag_name, stripped_input, parse_context)
+      klass.__send__(:new, tag_name, validated_link, parse_context)
     end
 
-    def self.validate_link!(link)
-      uri = URI.parse(link)
+    def self.validate_link(input:, retries: MAX_REDIRECTION_COUNT, method: Net::HTTP::Head)
+      uri = URI.parse(input.split.first)
       http = Net::HTTP.new(uri.host, uri.port)
       http.use_ssl = true if http.port == 443
 
-      req = Net::HTTP::Head.new(uri.request_uri)
+      req = method.new(uri.request_uri)
       req["User-Agent"] = "#{Settings::Community.community_name} (#{URL.url})"
       response = http.request(req)
 
       case response
       when Net::HTTPSuccess
-        response
+        input
       when Net::HTTPRedirection
-        warn "redirected to #{response['location']}"
+        raise StandardError, I18n.t("liquid_tags.unified_embed.tag.too_many_redirects") if retries.zero?
+
+        validate_link(input: response["location"], retries: retries - 1)
+      when Net::HTTPMethodNotAllowed
+        raise StandardError, I18n.t("liquid_tags.unified_embed.tag.invalid_url") if retries.zero?
+
+        validate_link(input: input, retries: retries, method: Net::HTTP::Get)
```

Comment:
With Dan's help, I came to understand that some endpoints do not allow `HEAD` requests, and will return the `NotAllowed` status. In those cases, we retry validation with a `GET` request.

Reply:
in fairness to you, that's probably a mistake in how they implemented their service. I've also seen plenty of systems where HEAD requests executed the full code for GET requests (wordpress is like this) but then discarded the body they carefully built along the way. 

Other people's code is the worst.